x-pack/plugin/inference/src/main/java/org/elasticsearch/xpack/inference/queries/SemanticQueryBuilder.java

```diff
@@ -0,0 +1,320 @@
+import org.elasticsearch.xpack.core.inference.results.SparseEmbeddingResults;
+import org.elasticsearch.xpack.inference.mapper.SemanticTextFieldMapper;
+
+import java.io.IOException;
+import java.util.Collection;
+import java.util.HashMap;
+import java.util.HashSet;
+import java.util.List;
+import java.util.Map;
+import java.util.Objects;
+import java.util.Set;
+
+import static org.elasticsearch.xpack.core.ClientHelper.ML_ORIGIN;
+import static org.elasticsearch.xpack.core.ClientHelper.executeAsyncWithOrigin;
+
+public class SemanticQueryBuilder extends AbstractQueryBuilder<SemanticQueryBuilder> {
+    public static final String NAME = "semantic_query";
+
+    private static final ParseField QUERY_FIELD = new ParseField("query");
+
+    private final String fieldName;
+    private final String query;
+
+    private SetOnce<InferenceServiceResults> inferenceResultsSupplier;
+    private InferenceServiceResults inferenceResults;
+
+    public SemanticQueryBuilder(String fieldName, String query) {
+        if (fieldName == null) {
+            throw new IllegalArgumentException("[" + NAME + "] requires a fieldName");
+        }
+        if (query == null) {
+            throw new IllegalArgumentException("[" + NAME + "] requires a " + QUERY_FIELD.getPreferredName() + " value");
+        }
+        this.fieldName = fieldName;
+        this.query = query;
+    }
+
+    public SemanticQueryBuilder(StreamInput in) throws IOException {
+        super(in);
+        this.fieldName = in.readString();
+        this.query = in.readString();
+        if (in.readBoolean()) {
+            inferenceResults = in.readNamedWriteable(InferenceServiceResults.class);
+            // The supplier is generally not used after the results are set, but set the supplier to maintain equality when serializing
+            // & deserializing
+            inferenceResultsSupplier = new SetOnce<>(inferenceResults);
+        }
+    }
+
+    private SemanticQueryBuilder(SemanticQueryBuilder other, SetOnce<InferenceServiceResults> inferenceResultsSupplier) {
+        this.fieldName = other.fieldName;
+        this.query = other.query;
+        this.boost = other.boost;
+        this.queryName = other.queryName;
+        this.inferenceResultsSupplier = inferenceResultsSupplier;
+    }
+
+    @Override
+    public String getWriteableName() {
+        return NAME;
+    }
+
+    @Override
+    public TransportVersion getMinimalSupportedVersion() {
+        return TransportVersions.SEMANTIC_TEXT_FIELD_ADDED;
+    }
+
+    @Override
+    protected void doWriteTo(StreamOutput out) throws IOException {
+        if (inferenceResultsSupplier != null && inferenceResults == null) {
+            throw new IllegalStateException("Inference results supplier is set, but inference results is null. Missing a rewriteAndFetch?");
+        }
+        out.writeString(fieldName);
+        out.writeString(query);
+        if (inferenceResults != null) {
```

Comment:
I could technically use `writeOptionalNamedWriteable()` here, but I can't use `in.readOptionalNamedWriteable` when reading the stream input, so it would result in an scenario where the `writeOptionalNamedWriteable()` implementation details would leak.